lib/utils/cli.go

```diff
@@ -49,65 +52,91 @@ const (
-	logrus.SetLevel(level)
+	logrus.SetLevel(logutils.SlogLevelToLogrusLevel(level))
 	switch purpose {
 	case LoggingForCLI:
 		// If debug logging was asked for on the CLI, then write logs to stderr.
 		// Otherwise, discard all logs.
-		if level == logrus.DebugLevel {
-			debugFormatter := logutils.NewDefaultTextFormatter(trace.IsTerminal(os.Stderr))
+		if level == slog.LevelDebug {
+			enableColors := trace.IsTerminal(os.Stderr)
+			debugFormatter := logutils.NewDefaultTextFormatter(enableColors)
 			_ = debugFormatter.CheckAndSetDefaults()
 			logrus.SetFormatter(debugFormatter)
-			logrus.SetOutput(os.Stderr)
+
+			w := logutils.NewSharedWriter(os.Stderr)
+			logrus.SetOutput(w)
+			slog.SetDefault(slog.New(logutils.NewSlogTextHandler(w, &logutils.SlogTextHandlerConfig{
+				Level:        level,
+				EnableColors: enableColors,
+				WithCaller:   true,
+			})))
+			return
 		} else {
+			slog.SetDefault(slog.New(slog.NewTextHandler(io.Discard, nil)))
 			logrus.SetOutput(io.Discard)
 		}
 	case LoggingForDaemon:
-		logrus.SetFormatter(logutils.NewDefaultTextFormatter(trace.IsTerminal(os.Stderr)))
-		logrus.SetOutput(os.Stderr)
+		enableColors := trace.IsTerminal(os.Stderr)
+		w := logutils.NewSharedWriter(os.Stderr)
+
+		logrus.SetFormatter(logutils.NewDefaultTextFormatter(enableColors))
+		logrus.SetOutput(w)
+		slog.SetDefault(slog.New(logutils.NewSlogTextHandler(w, &logutils.SlogTextHandlerConfig{
+			Level:        level,
+			EnableColors: enableColors,
+			WithCaller:   true,
+		})))
 	}
 }
 
+var initTestLoggerOnce = sync.Once{}
+
 // InitLoggerForTests initializes the standard logger for tests.
 func InitLoggerForTests() {
-	// Parse flags to check testing.Verbose().
-	flag.Parse()
-
-	logger := logrus.StandardLogger()
-	logger.ReplaceHooks(make(logrus.LevelHooks))
-	logrus.SetFormatter(logutils.NewTestJSONFormatter())
-	logger.SetLevel(logrus.DebugLevel)
-	logger.SetOutput(os.Stderr)
-	if testing.Verbose() {
-		return
-	}
-	logger.SetLevel(logrus.WarnLevel)
-	logger.SetOutput(io.Discard)
+	initTestLoggerOnce.Do(func() {
+		// Parse flags to check testing.Verbose().
+		flag.Parse()
+
+		level := slog.LevelWarn
+		w := io.Discard
+		if testing.Verbose() {
+			level = slog.LevelDebug
+			w = os.Stderr
+		}
+
+		logger := logrus.StandardLogger()
+		logger.SetFormatter(logutils.NewTestJSONFormatter())
+		logger.SetLevel(logutils.SlogLevelToLogrusLevel(level))
+
+		output := logutils.NewSharedWriter(w)
+		logger.SetOutput(output)
+		slog.SetDefault(slog.New(logutils.NewSlogJSONHandler(output, level)))
+	})
 }
 
-// NewLoggerForTests creates a new logger for test environment
+// NewLoggerForTests creates a new logrus logger for test environments.
 func NewLoggerForTests() *logrus.Logger {
-	logger := logrus.New()
-	logger.ReplaceHooks(make(logrus.LevelHooks))
-	logger.SetFormatter(logutils.NewTestJSONFormatter())
-	logger.SetLevel(logrus.DebugLevel)
-	logger.SetOutput(os.Stderr)
-	return logger
+	InitLoggerForTests()
```

Comment:
We ought to be careful w/ this change in behaviour - I know previously I've considering invoking `NewLoggerForTests` with the specific intention of having a new instance of a logger and not the global logger. If we do change the behaviour, we should change the godoc to indicate this now returns the global logger.

Reply:
Whilst I wouldn't encourage it, I can imagine a scenario where someone wishes to capture the outputs of the logging and assert on these during the test. In this situation, they'd call `NewLoggerForTests` and then call `.SetOutput` on this logger - with this change, they'd then be changing the Output for the global logger. 

I suppose really this was my original interpretation - when I'd originally seen we had a NewLoggerForTests and InitLoggerForTests a few months ago, this was the explanation for the existence of both that I'd come to with the little context I had. The `New` would lead me to believe that this wasn't the global logger.